src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs

```diff
@@ -0,0 +1,111 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Test.Common;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Http.Functional.Tests
+{
+    using Configuration = System.Net.Test.Common.Configuration;
+
+    public abstract class HttpClientHandlerTest_Headers : HttpClientHandlerTestBase
+    {
+        public HttpClientHandlerTest_Headers(ITestOutputHelper output) : base(output) { }
+
+        [Fact]
+        public async Task SendAsync_UserAgent_CorrectlyWritten()
+        {
+            string userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.18 Safari/537.36";
+
+            await LoopbackServerFactory.CreateClientAndServerAsync(async uri =>
+            {
+                using (var client = CreateHttpClient())
+                {
+                    var message = new HttpRequestMessage(HttpMethod.Get, uri);
+                    message.Headers.TryAddWithoutValidation("User-Agent", userAgent);
+                    (await client.SendAsync(message).ConfigureAwait(false)).Dispose();
+                }
+            },
+            async server =>
+            {
+                HttpRequestData requestData = await server.HandleRequestAsync(HttpStatusCode.OK);
+
+                string agent = requestData.GetSingleHeaderValue("User-Agent");
+                Assert.Equal(userAgent, agent);
+            });
+        }
+
+        [Fact]
+        public async Task SendAsync_SpecialCharacterHeader_Success()
+        {
+            string headerValue = "header name with underscore";
+            await LoopbackServerFactory.CreateClientAndServerAsync(async uri =>
+            {
+                using (var client = CreateHttpClient())
+                {
+                    var message = new HttpRequestMessage(HttpMethod.Get, uri);
+                    message.Headers.TryAddWithoutValidation("x-Special_name", "header name with underscore");
+                    (await client.SendAsync(message).ConfigureAwait(false)).Dispose();
+                }
+            },
+            async server =>
+            {
+                HttpRequestData requestData = await server.HandleRequestAsync(HttpStatusCode.OK);
+
+                string header = requestData.GetSingleHeaderValue("x-Special_name");
+                Assert.Equal(header, headerValue);
+            });
+        }
+
+        [OuterLoop("Uses external server")]
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task SendAsync_GetWithValidHostHeader_Success(bool withPort)
+        {
+            var m = new HttpRequestMessage(HttpMethod.Get, Configuration.Http.SecureRemoteEchoServer);
+            m.Headers.Host = withPort ? Configuration.Http.SecureHost + ":123" : Configuration.Http.SecureHost;
```

Comment:
Modifying the Host header like this (adding a random port number) will probably make this test fail since it is talking to a real external server.

Reply:
Maybe this should be "443" ??

Reply:
updated. This was existing test I moved to new class file.